src/plugins/camera/camera_impl.cpp

```diff
@@ -899,39 +899,59 @@ void CameraImpl::process_camera_information(const mavlink_message_t& message)
-            const auto temp_callback = _information.subscription_callback;
-            const auto temp_information = _information.data;
-            _parent->call_user_callback(
-                [temp_callback, temp_information]() { temp_callback(temp_information); });
-        }
+    std::lock_guard<std::mutex> lock(_information.mutex);
+
+    _information.data.vendor_name = (char*)(camera_information.vendor_name);
+    _information.data.model_name = (char*)(camera_information.model_name);
+    _information.data.focal_length_mm = camera_information.focal_length;
+    _information.data.horizontal_sensor_size_mm = camera_information.sensor_size_h;
+    _information.data.vertical_sensor_size_mm = camera_information.sensor_size_v;
+    _information.data.horizontal_resolution_px = camera_information.resolution_h;
+    _information.data.vertical_resolution_px = camera_information.resolution_v;
+
+    if (_information.subscription_callback) {
+        const auto temp_callback = _information.subscription_callback;
+        const auto temp_information = _information.data;
+        _parent->call_user_callback(
+            [temp_callback, temp_information]() { temp_callback(temp_information); });
     }
 
-    if (!_camera_definition) {
-        std::string content{};
-        auto succeeded = fetch_camera_definition(camera_information, content);
-
-        if (succeeded) {
-            _camera_definition.reset(new CameraDefinition());
-            _camera_definition->load_string(content);
-            refresh_params();
-            LogDebug() << "Successfully loaded camera definition";
-        } else {
-            LogDebug() << "Failed to fetch camera definition!";
-        }
+    if (should_fetch_camera_definition()) {
+        _is_fetching_camera_definition = true;
+
+        std::thread([this, camera_information]() {
+            std::string content{};
+            auto succeeded = fetch_camera_definition(camera_information, content);
+
+            if (succeeded) {
+                LogDebug() << "Successfully loaded camera definition";
+
+                _camera_definition.reset(new CameraDefinition());
+                _camera_definition->load_string(content);
+                refresh_params();
+            } else {
+                LogDebug() << "Failed to fetch camera definition!";
+
+                if (++_camera_definition_fetch_count >= 3) {
+                    LogWarn() << "Failed to fetch camera definition 3 times, giving up";
+
+                    std::lock_guard<std::mutex> thread_lock(_information.mutex);
+                    _has_camera_definition_timed_out = true;
+                }
+            }
+
+            std::lock_guard<std::mutex> thread_lock(_information.mutex);
+            _is_fetching_camera_definition = false;
+        }).detach();
     }
 }
 
+bool CameraImpl::should_fetch_camera_definition() const
+{
+    return !_camera_definition && !_is_fetching_camera_definition &&
+           !_has_camera_definition_timed_out;
```

Comment:
So if we ever timeout we won't try ever again? Shouldn't we try again after a longer timeout or so?

Reply:
Not sure... what would be the use-case for that? It takes 3-5 seconds to timeout, so that's 15 before it gives up. Two questions:

1. When does it happen that it fails 3 times in a row when you connect the app, but works later?
2. If it retries every 30 or 60 seconds, my guess is that the user will learn how to restart the app faster than that.

My feeling is that if you need the camera controls (e.g. for a survey), you will never want to get them in the middle of the mission: you want them before you start the mission, as a preflight check.

But happy to make it retry every 30 or 60 seconds if you think there is a use-case for that :+1:

Reply:
Yer, you're probably right. I was just thinking of the case where the companion takes some time to boot up.

Reply:
Is that an issue? The `CAMERA_INFORMATION` message arrives only when the camera component sends it, and at this point it should be ready to serve the camera definition, shouldn't it?

Reply:
Oh, you're right. All good then.